datasets/plugins/batch/batch_dataset_plugin.py

```diff
@@ -0,0 +1,173 @@
+
+        if not isinstance(data, pd.DataFrame):
+            assert ValueError("data is not a pandas DataFrame")
+
+        if self.partition_by:
+            partition_cols = self.partition_by.split(",")
+        else:
+            partition_cols = list()
+
+        if self.path is None or "run_id" in partition_cols:
+            # Only partition on run_id if @dataset(path="s3://..") is not given
+            # or run_id is in partition_cols
+            if "run_id" not in partition_cols:
+                partition_cols.append("run_id")
+            self.run_id = self._executor.current_run_id  # DO NOT ALLOW OVERWRITE OF ANOTHER RUN ID
+            data["run_id"] = self.run_id
+
+        data.to_parquet(
+            self._get_dataset_path(),
+            engine="pyarrow",
+            compression="snappy",
+            index=False,
+            partition_cols=partition_cols,
+            **kwargs,
+        )
+
+    def read_dask(
+        self, columns: Optional[str] = None, run_id: Optional[str] = None, **kwargs
+    ) -> "dd.DataFrame":
+        import dask.dataframe as dd
+
+        path, filters, read_columns = self._get_path_filters_columns(columns, run_id=run_id)
+        return dd.read_parquet(
+            path,
+            columns=read_columns,
+            filters=filters if filters and len(filters) else None,
+            engine="pyarrow",
+            **kwargs,
+        )
+
+    def read_spark(
+        self, columns: Optional[str] = None, run_id: Optional[str] = None, conf=None, **kwargs
+    ) -> "pyspark.sql.DataFrame":
+        from pyspark import SparkConf
+        from pyspark.sql import DataFrame, SparkSession
+
+        path, _, read_columns = self._get_path_filters_columns(columns, run_id=run_id)
+
+        read_columns = read_columns if read_columns else ["*"]
+        if self.run_id:
+            read_columns.append("run_id")
+
+        if conf is None:
+            conf = SparkConf()
+        spark_session: SparkSession = SparkSession.builder.config(conf=conf).getOrCreate()
+
+        df: DataFrame = spark_session.read.parquet(path).select(*read_columns)
+        if self.run_id:
+            df = df.where(df["run_id"] == self.run_id)
+        return df
+
+    @classmethod
+    def _register_dataset_path_func(cls, func: callable):
+        cls._dataset_path_func = func
+
+    def _get_dataset_path(self) -> str:
+        if self.path is not None:
+            return self.path
+        else:
+            if BatchDatasetPlugin._dataset_path_func:
+                return BatchDatasetPlugin._dataset_path_func(self)
+            else:
+                return str(
+                    Path(self._executor.datastore_path)
+                    / "datastore"
+                    / (self.program_name if self.program_name else self._executor.current_program_name)
+                    / self.name
+                )
+
+    def __str__(self):
+        return self.__repr__()
```

Comment:
Shouldn't need this `__str__` method, seems like it's just duplicating the base class implementation.